net/graphsync_fetcher.go

```diff
@@ -0,0 +1,180 @@
+// - a future update sync that is likely to fetch only a few
+//
+// To do this, the Graphsync fetcher starts fetching a single tipset at a time,
+// then gradually ramps up to fetch lots of tipsets at once, up to a fixed limit
+//
+// The constants below determine the maximum number of tipsets fetched at once
+// (maxRecursionDepth) and how fast the ramp up is (recursionMultipler)
+const maxRecursionDepth = 64
+const recursionMultiplier = 4
+
+// FetchTipSets gets Tipsets starting from the given tipset key and continuing till
+// the done function returns true or errors
+//
+// For now FetchTipSets operates in two parts:
+// 1. It fetches relevant blocks through Graphsync, which writes them to the block store
+// 2. It reads them from the block store and validates there syntax as blocks
+// and constructs a tipset
+// This does have a potentially unwanted side effect of writing blocks to the block store
+// that later don't validate (bitswap actually does this as well)
+// In the future, the blocks will be validated directly through graphsync as
+// go-filecoin migrates to the same IPLD library used by go-graphsync (go-ipld-prime)
+func (gsf *GraphSyncFetcher) FetchTipSets(ctx context.Context, tsKey types.TipSetKey, from peer.ID, done func(types.TipSet) (bool, error)) ([]types.TipSet, error) {
+	cids := tsKey.ToSlice()
+	err := gsf.fetchBlocks(ctx, cids, from)
+	if err != nil {
+		return nil, err
+	}
+	startingTipset, err := gsf.loadTipsetFromCids(ctx, cids)
+	if err != nil {
+		return nil, err
+	}
+	out := []types.TipSet{startingTipset}
+	isDone, err := done(startingTipset)
+	if err != nil {
+		return nil, err
+	}
+	recursionDepth := 1
+	ts := startingTipset
+	for !isDone {
+		err := gsf.fetchBlocksRecursively(ctx, ts.At(0).Cid(), from, recursionDepth)
+		if err != nil {
+			return nil, err
+		}
+		for i := 0; i < recursionDepth; i++ {
+			tsKey, err := ts.Parents()
+			if err != nil {
+				return nil, err
+			}
+			cids := tsKey.ToSlice()
+			ts, err = gsf.loadTipsetFromCids(ctx, cids)
+			if err != nil {
+				return nil, err
+			}
+			out = append(out, ts)
+			isDone, err = done(ts)
+			if err != nil {
+				return nil, err
+			}
+			if isDone {
+				break
+			}
+		}
+		if recursionDepth < maxRecursionDepth {
+			recursionDepth *= recursionMultiplier
+		}
+	}
+	return out, nil
+}
+
+// fetchBlocks requests a single set of cids as individual bocks, fetching
+// non-recursively
+func (gsf *GraphSyncFetcher) fetchBlocks(ctx context.Context, cids []cid.Cid, from peer.ID) error {
+	selector := gsf.ssb.Matcher().Node()
+	errChans := make([]<-chan error, 0, len(cids))
+	for _, c := range cids {
+		_, errChan := gsf.exchange.Request(ctx, from, cidlink.Link{Cid: c}, selector)
+		errChans = append(errChans, errChan)
+	}
+	for _, errChan := range errChans {
+		for err := range errChan {
+			return err
+		}
+	}
+	return nil
+}
+
+// fetchBlocksRecursively gets multiple sets of parent blocks starting from a baseCid, up to
```

Comment:
More idiomatically "...gets the blocks from recursionDepth ancestor tipsets starting from baseCid..."